src/System.Net.Primitives/src/System/Net/Sockets/SocketAddress.cs

```diff
@@ -0,0 +1,283 @@
+            else
+            {
+                throw new SocketException(SocketError.AddressFamilyNotSupported);
+            }
+        }
+
+        internal IPEndPoint GetIPEndPoint()
+        {
+            IPAddress address = GetIPAddress();
+            int port = (int)((Buffer[2] << 8 & 0xFF00) | (Buffer[3]));
+            return new IPEndPoint(address, port);
+        }
+
+        // For ReceiveFrom we need to pin address size, using reserved Buffer space
+        internal void CopyAddressSizeIntoBuffer()
+        {
+            Buffer[Buffer.Length - IntPtr.Size] = unchecked((byte)(InternalSize));
+            Buffer[Buffer.Length - IntPtr.Size + 1] = unchecked((byte)(InternalSize >> 8));
+            Buffer[Buffer.Length - IntPtr.Size + 2] = unchecked((byte)(InternalSize >> 16));
+            Buffer[Buffer.Length - IntPtr.Size + 3] = unchecked((byte)(InternalSize >> 24));
+        }
+
+        // Can be called after the above method did work
+        internal int GetAddressSizeOffset()
+        {
+            return Buffer.Length - IntPtr.Size;
+        }
+
+        // For ReceiveFrom we need to update the address size upon IO return
+        internal unsafe void SetSize(IntPtr ptr)
+        {
+            // Apparently it must be less or equal the original value since ReceiveFrom cannot reallocate the address buffer
+            InternalSize = *(int*)ptr;
+        }
+        public override bool Equals(object comparand)
+        {
+            SocketAddress castedComparand = comparand as SocketAddress;
+            if (castedComparand == null || this.Size != castedComparand.Size)
+            {
+                return false;
+            }
+            for (int i = 0; i < this.Size; i++)
+            {
+                if (this[i] != castedComparand[i])
+                {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        public override int GetHashCode()
+        {
+            if (_changed)
+            {
+                _changed = false;
+                _hash = 0;
+
+                int i;
+                int size = Size & ~3;
+
+                for (i = 0; i < size; i += 4)
+                {
+                    _hash ^= (int)Buffer[i]
+                            | ((int)Buffer[i + 1] << 8)
+                            | ((int)Buffer[i + 2] << 16)
+                            | ((int)Buffer[i + 3] << 24);
+                }
+                if ((Size & 3) != 0)
+                {
+                    int remnant = 0;
+                    int shift = 0;
+
+                    for (; i < Size; ++i)
+                    {
+                        remnant |= ((int)Buffer[i]) << shift;
+                        shift += 8;
+                    }
+                    _hash ^= remnant;
+                }
+            }
+            return _hash;
+        }
+
+        public override string ToString()
+        {
+            StringBuilder bytes = new StringBuilder();
```

Comment:
There's enough StringBuilder usage (or potential for StringBuilder usage) that it does seem like it could be valuable to bring in StringBuilderCache.